allennlp/data/semparse/type_declarations/type_declaration.py

```diff
@@ -474,3 +474,25 @@ def get_valid_actions(name_mapping: Dict[str, str],
 
 
 START_TYPE = NamedBasicType(START_SYMBOL)
+
+# TODO(mattg): We're hard-coding three lambda variables here.  This isn't a great way to do
+# this; it's just something that works for now, that we can fix later if / when it's needed.
+# If you allow for more than three nested lambdas, or if you want to use different lambda
+# variable names, you'll have to change this somehow.
+LAMBDA_VARIABLES = set(['x', 'y', 'z'])
+
+def is_nonterminal(production: str) -> bool:
+    # TODO(pradeep): This is pretty specific to the assumptions made in converting types to
```

Comment:
I don't think it should be world-specific. The type declaration should define the logic for distinguishing terminals from non-terminals, because that is part of the type declaration. I guess your point is that the world can potentially add terminals. But since we know what their types are, it shouldn't be hard for the type declaration to define this logic based on types, instead of the surface forms of the productions. I added a TODO for this.

Reply:
Sorry, you're right, I meant the type declaration corresponding to the world.  Just not the base type declaration.